crates/ruff_linter/src/rules/refurb/rules/slice_to_remove_prefix_or_suffix.rs

```diff
@@ -0,0 +1,498 @@
+    //     text = f"{prefix} something completely different"
+    //     text = text[len(prefix):]
+    // ```
+    if body.len() != 1 {
+        return None;
+    };
+
+    // Variable names correspond to:
+    // ```python
+    // if test:
+    //     else_or_target_name = value[slice]
+    // ```
+    let ast::StmtAssign {
+        value,
+        targets,
+        range: _,
+    } = body[0].as_assign_stmt()?;
+    if targets.len() != 1 {
+        return None;
+    };
+    let else_or_target_name = &targets[0].as_name_expr()?.id;
+    let ast::ExprSubscript { value, slice, .. } = value.as_subscript_expr()?;
+
+    affix_removal_data_core_logic(value, test, else_or_target_name, slice)
+}
+
+/// Suppose given a statement of the form:
+/// ```python
+/// if test:
+///     else_or_target_name = value[slice]
+/// ```
+/// or an expression of the form:
+/// ```python
+/// value[slice] if test else else_or_target_name
+/// ```
+/// This function verifies that
+///     - `value` and `else_or_target_name`
+/// are equal to a common name `text`
+///     - `test` is of the form `text.startswith(prefix)`
+/// or `text.endswith(suffix)`
+///     - `slice` has no upper bound in the case of a prefix,
+/// and no lower bound in the case of a suffix
+///
+/// If these conditions are satisfied, the function
+/// returns the corresponding `RemoveAffixData` object;
+/// otherwise it returns `None`.
+fn affix_removal_data_core_logic<'a>(
+    value: &'a ast::Expr,
+    test: &'a ast::Expr,
+    else_or_target_name: &'a ast::name::Name,
+    slice: &'a ast::Expr,
+) -> Option<RemoveAffixData<'a>> {
+    let body_name = &value.as_name_expr()?.id;
+    let slice = slice.as_slice_expr()?;
+    let test_name = &test
+        .as_call_expr()?
+        .func
+        .as_attribute_expr()?
+        .value
+        .as_name_expr()?
+        .id;
+    let func_name = test
+        .as_call_expr()?
+        .func
+        .as_attribute_expr()?
+        .attr
+        .id
+        .as_str();
+
+    let func_args = &test.as_call_expr()?.arguments.args;
+    if func_args.len() != 1 {
+        return None;
+    }
+
+    let affix = &func_args[0];
+    match (body_name, test_name, else_or_target_name, slice, func_name) {
+        (text, a, b, _, "startswith") if text == a && a == b && slice.upper.is_none() => {
+            Some(RemoveAffixData {
+                text,
+                bound: slice.lower.as_ref()?,
+                affix_query: AffixQuery {
+                    kind: AffixKind::StartsWith,
+                    affix,
+                },
+            })
+        }
+        (text, a, b, _, "endswith") if text == a && a == b && slice.lower.is_none() => {
+            Some(RemoveAffixData {
+                text,
+                bound: slice.upper.as_ref()?,
+                affix_query: AffixQuery {
+                    kind: AffixKind::EndsWith,
+                    affix,
+                },
+            })
+        }
+        _ => None,
+    }
```

Comment:
A little less nested:

```suggestion
    if body_name != test_name || test_name != else_or_target_name {
        return None;
    }

    let (affix_kind, bound) = match func_name {
        "startswith" if slice.upper.is_none() => (AffixKind::StartsWith, slice.lower.as_ref()?),
        "endswith" if slice.lower.is_none() => (AffixKind::EndsWith, slice.upper.as_ref()?),
        _ => return None,
    };

    Some(RemoveAffixData {
        text: body_name,
        bound,
        affix_query: AffixQuery {
            kind: affix_kind,
            affix,
        },
    })
```

Reply:
Done (though it looks a little different because I implemented one of your other suggestions as well)